web-api/src/personalization/rerank.rs

```diff
@@ -94,34 +82,28 @@ pub(super) fn rerank_by_scores(
 ) {
     let interest_scores = rerank_by_interest(coi_system, documents, interests, time);
     let tag_weight_scores = rerank_by_tag_weight(documents, tag_weights);
-    let mut elasticsearch_scores = HashMap::with_capacity(documents.len());
 
-    for document in documents.iter_mut() {
-        elasticsearch_scores.insert(document.id.clone(), document.score);
-        document.score = score_weights[0] * interest_scores[&document.id]
-            + score_weights[1] * tag_weight_scores[&document.id]
-            + score_weights[2] * document.score;
+    let scores = documents
+        .iter()
+        .map(|document| {
+            let interest_score = interest_scores
+                .get(&document.id)
+                .copied()
+                .unwrap_or_default();
+            let tag_weight_score = tag_weight_scores
+                .get(&document.id)
+                .copied()
+                .unwrap_or_default();
+            score_weights[0] * interest_score
+                + score_weights[1] * tag_weight_score
+                + score_weights[2] * document.score
+        })
+        .collect_vec();
+    for (document, score) in documents.iter_mut().zip(scores) {
+        document.score = score;
```

Comment:
:wrench:  There is no reasons to make this two loops and an additional collect if we anyway look up the scores by document id. We can just compute the compound score in the single place we use it.

Reply:
the main difference is that we don't clone all the ids multiple times anymore. i discussed this with @acrrd and he also agreed that collecting one additional time instead of cloning multiple times is a reasonable tradeoff. if there are better ways to handle this i'm happy to change it, but usually the borrow checker complains about mutable and immutable borrows at the same time.

Reply:
:+1: I missed that it requires a cloning before the loop and only checked that using the loop doesn't cause any additional cloning by itself.